aeneas/src/mach/GlobalRegAlloc.v3

```diff
@@ -0,0 +1,603 @@
+		if (colorVerbose) printColor();
+		allocateFromColor();
+	}
+	def computeLiveoutForBlock(block: SsaBlock) {
+		liveness.clearRow(liveoutRow);
+		for (e in block.succs()) {
+			var succ = e.dest;
+			// union the succesor's live-in set into this.
+			liveness.or(liveoutRow, succ.mark);
+		}
+	}
+	def allocateFromColor() {
+		cur = codegen.last;
+		while (cur != null) {
+			if (verbose) Terminal.put(buf.putInstrV(cur).toString());
+			pos++;
+			prev = cur.prev;
+			next = cur.next;
+			var opcode = cur.opcode();
+			if (opcode == ArchInstrs.ARCH_BLOCK_END) {
+				release();
+				var block = Operand.Label.!(cur.operands[0]).block;
+				computeLiveoutForBlock(block);
+				for (i < vars.length) {
+					if (liveness[liveoutRow, i]) updateRegState(vars[i]);
+				}
+				cur = prev;
+				continue;
+			}
+			
+			var operands = cur.operands;
+			if (opcode == ArchInstrs.ARCH_PARMOVE) { // perform parallel move
+				parmoveDests.resize(0);
+				codegen.gatherParallelMoveDests(cur, parmoveDests);
+				for (j < parmoveDests.length) recursivelyEmitMoves(j, next);
+				for (i < operands.length) match (operands[i]) {
+					Def(vreg, constraint) => {
+						var gReg = coloring[vreg.varNum];
+						freeReg(gReg);
+						if (regSet.isReg(gReg)) codegen.insertRestoreLocal(vreg, gReg, next);
+					}
+					Use(vreg, assignment) => {
+						if (!regSet.isStack(assignment) || !vreg.isConst()) {
+							updateRegState(vreg);
+						}
+					}
+					_ => ;
+				}
+				cur.remove();
+				if (verbose) dumpState("");
+				cur = prev;
+				continue;
+			}
+			moves.clear();
+			movesPrev.clear();
+			for (i < operands.length) match (operands[i]) {
+				Def(vreg, constraint) => {
+					var loc = allocDefReg(vreg, constraint);
+					operands[i] = Operand.Def(vreg, loc);
+				}
+				Overwrite(vreg, use, constraint)    => {
+					var loc = allocDefReg(vreg, constraint), newReg = byte.!(loc);
+					var gReg = coloring[use.varNum];
+					if (newReg != gReg) {
+						if (regSet.isReg(gReg)) movesPrev.addRegMove(use, gReg, newReg);
+						else movesPrev.reloads.put(use, newReg);
+					}
+					regUsage[loc] = pos;
+					operands[i] = Operand.Overwrite(vreg, use, loc);
+				}
+				Use(vreg, constraint) => {
+					var loc = allocUseReg(vreg, constraint);
+					operands[i] = Operand.Use(vreg, loc);
+				}
+				Kill(constraint) => {
+					if (constraint < regSet.regSets.length) {
+						for (reg in regSet.regSets[constraint]) {
+							spillReg(reg);
+						}
+					}
+				}
+				_ => ;
+			}
+			for (i < operands.length) match (operands[i]) {
+				Overwrite(vreg, use, assignment)    => {
+					if (!regSet.isStack(assignment) || !use.isConst()) {
+						updateRegState(use);
+					}
+				}
+				Use(vreg, assignment) => {
+					if (!regSet.isStack(assignment) || !vreg.isConst()) {
+						updateRegState(vreg);
+					}
+				}
+				_ => ;
+			}
+			if (verbose) dumpState("");
+			moves.emit(codegen, cur.next);
+			movesPrev.emit(codegen, cur);
+			if (opcode == ArchInstrs.ARCH_ENTRY) {
```

Comment:
I had another idea to fix the merging of live-in sets from successors (in the interference-graph building phase): make constants interfere at `ARCH_ENTRY`. That may be cheaper overall.

Reply:
In this case, the constant might not interfere with other variable (not defined at `ARCH_ENTRY`, but defined on the path from entry to using constant).